### PR TITLE
KNPC cleanup and optimizations

### PR DESCRIPTION
**Pull request by TheFakeElon**
## About The Pull Request

KNPCs should be more performant now, not that they were very expensive to begin with though, relatively speaking.

## Changelog
:cl:
code: KNPC optimizations
/:cl:

### CODE REVIEW

File: nsv13/code/modules/overmap/knpc.dm [hunk 1]
```diff
@@ -253,35 +241,33 @@ GLOBAL_LIST_EMPTY(knpcs)
...
 		if(H.faction_check_mob(M))
 			continue
-		if(!can_see(H, M, HA.view_range))
-			return
 		. += M
 	//Check for nearby mechas....
-	if(GLOB.mechas_list?.len)
-		for(var/obj/mecha/OM in GLOB.mechas_list)
+	if(length(GLOB.mechas_list))
+		for(var/obj/mecha/OM as() in GLOB.mechas_list)
```
Comment (4310V343k) [inline nsv13/code/modules/overmap/knpc.dm:257]: `as anything` just looks better

File: nsv13/code/modules/overmap/knpc.dm [hunk 2]
```diff
@@ -311,16 +297,16 @@ This is to account for sec Ju-Jitsuing boarding commandos.
 	if(!..())
 		return 0
 	var/mob/living/carbon/human/H = HA.parent
-	var/obj/A = H.get_active_held_item()
+	var/obj/item/gun/G = H.get_active_held_item()
 	//We already have a gun
-	if(A && istype(A, /obj/item/gun))
+	if(G && istype(G))
 		return 0
```
Comment (4310V343k) [inline nsv13/code/modules/overmap/knpc.dm:303]: I mean

**Review by 4310V343k — COMMENTED** [review]
i know i'm useful

File: nsv13/code/modules/overmap/knpc.dm [hunk 3]
```diff
@@ -35,8 +35,8 @@ GLOBAL_LIST_EMPTY(knpcs)
 
 /mob/living/carbon/human/ai_boarder/Initialize()
 	. = ..()
-	var/datum/outfit/P = pick(outfit)
-	var/datum/outfit/O = new P
+	var/outfit_path = pick(outfit)
```
[the remark this answers is not in the document]
Comment (TheFakeElon) [reply nsv13/code/modules/overmap/knpc.dm:38]: Yep! It refuses to compile if you use pick() to specify type. I don't think this is a problem on newer BYOND versions but we're still running an outdated version, so this is needed

File: nsv13/code/modules/overmap/knpc.dm (hunk 1 above)
Comment (TheFakeElon) [reply nsv13/code/modules/overmap/knpc.dm:257]: `as anything` is needlessly longer, as() is generally more common on most codebases but ultimately it's down to preference.

File: nsv13/code/modules/overmap/knpc.dm (hunk 2 above)
Comment (TheFakeElon) [reply nsv13/code/modules/overmap/knpc.dm:303]: There's literally nothing wrong with this. At worst it's cleaner and at best it's faster, depending on DM's compiler side optimizations

Comment (4310V343k) [reply nsv13/code/modules/overmap/knpc.dm:303]: always look at the last line of code
```suggestion
		return FALSE
```
it's just cosmetic
but it's everywhere

File: nsv13/code/modules/overmap/knpc.dm (hunk 1 above)
Comment (4310V343k) [reply nsv13/code/modules/overmap/knpc.dm:257]: as() is less readable

File: nsv13/code/modules/overmap/knpc.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (4310V343k) [reply nsv13/code/modules/overmap/knpc.dm:38]: lame

File: nsv13/code/modules/overmap/knpc.dm (hunk 1 above)
Comment (TheFakeElon) [reply nsv13/code/modules/overmap/knpc.dm:257]: How?

File: nsv13/code/modules/overmap/knpc.dm (hunk 2 above)
Comment (TheFakeElon) [reply nsv13/code/modules/overmap/knpc.dm:303]: I didn't even change that line, I'd rather keep things consistent with the rest of the file than just changing one random line. Feel free to make a PR addressing the whole file though, it's just out of scope for this PR

File: nsv13/code/modules/overmap/knpc.dm [hunk 4]
```diff
@@ -67,114 +67,112 @@ GLOBAL_LIST_EMPTY(knpcs)
...
-		var/obj/machinery/door/firedoor/border_only/fuckingMonsterMos = locate(/obj/machinery/door/firedoor/border_only) in get_turf(get_step(H, H.dir))
-		if(fuckingMonsterMos && istype(fuckingMonsterMos))
-			fuckingMonsterMos.open()
+	var/turf/targturf = get_turf(target)
+	if(dest && dest == targturf || H.incapacitated())
+		return FALSE //No need to recalculate the path.
+	dest = targturf
+	var/turf/T = get_step(H, H.dir)
+	var/obj/structure/dense_object = locate(/obj/structure) in T //If we're stuck
+	if(climbable[dense_object.type])
```
Comment (covertcorvid) [inline nsv13/code/modules/overmap/knpc.dm:93]: getting a runtime here in testing - cannot read null.type
none of them will move anywhere.

**Comment by Bokkiewokkie** [comment]
Why is all of the past 2 months of PR's in here now and github says there's 113 changed files

**Comment by TheFakeElon** [comment]
The germans

**Review by Bokkiewokkie — APPROVED** [review]

**Pull request merged**